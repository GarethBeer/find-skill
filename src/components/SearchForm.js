import React, { useEffect, useRef, useState } from "react";
import PropTypes from "prop-types";
import "../styles/SearchForm.css";

let autoComplete;

const SearchForm = ({ className, title, onFill, onPick }) => {
  const ref = useRef();
  const [selectedOption, setSelectedOption] = useState();

  const onValueChange = (e) => {
    setSelectedOption(Number(e.target.value));
  };

  // console.log("sopt", selectedOption);

  useEffect(() => {
    const onScriptLoad = async (input) => {
      autoComplete = await new window.google.maps.places.Autocomplete(
        ref.current,
        input
      );
      autoComplete.setComponentRestrictions({
        country: "gb",
      });
      autoComplete.setTypes(["(cities)"]);

      autoComplete.setFields(["geometry"]);
      autoComplete.addListener("place_changed", () => {
        const place = autoComplete.getPlace();

        const latititude = place.geometry.location.lat();
        const longitude = place.geometry.location.lng();
        const latLng = { lat: latititude, lng: longitude };

        onFill(latLng);
      });
    };

    onScriptLoad();
  }, [onFill]);

  return (
    <div className="searchform-container">
      <h4>Pick your town</h4>
      <input
        ref={ref}
        placeholder="Enter a City"
        title={title}
        className={className}
      />
      <h4>Pick the search radius</h4>
      <form className="radius-container" onSubmit={onPick(selectedOption || 8)}>
        <label className="radio" htmlFor="8">
          <input
            id="8"
            type="radio"
            value="8"
            checked={selectedOption === 8}
            onChange={onValueChange}
          />
          Large
        </label>

        <label className="radio" htmlFor="9">
          <input
            id="9"
            type="radio"
            value="9"
            checked={selectedOption === 9}
            onChange={onValueChange}
          />
          Medium
        </label>

        <label className="radio" htmlFor="12">
          <input
            id="12"
            type="radio"
            value="12"
            checked={selectedOption === 12}
            onChange={onValueChange}
          />
          Small
        </label>
      </form>
    </div>
  );
};

SearchForm.propTypes = {
  className: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  onFill: PropTypes.func.isRequired,
  onPick: PropTypes.func.isRequired,
};

export default SearchForm;
