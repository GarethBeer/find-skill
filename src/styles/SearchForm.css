.searchform-container {
    display: flex;
    flex-direction: column;
    border: solid 1px rgb(5, 107, 5);
    padding:2px;
}
input {
    margin: 0vw 1vw;
    border-radius: 4px;
    padding: 3px;
    margin-bottom: 1.5vh;
}
.radius-container {
    display: flex;
    justify-content: space-evenly;
}
.radio {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    margin-left: 1vw;
    cursor: pointer;
    font-size: 2vh;
    align-items: center;
}
.radio input {
   cursor: pointer;
    height: 3vh;
    width: 3vh;
}
.radio input:checked ~ .checkmark {
    background-color: rgb(5, 107, 5);
}
.radio input:checked {
    display: block;
}
.radio:after {
    top: 3px;
    left: 3px;
    width: 6px;
    height: 6px;
    border-radius: 50%;
    background: white;
    color: lawngreen;
}
@media only screen and (max-width: 732px) {
    .searchform-container {
        width: 48%;
    }
    input {
        padding: 7px;
    }
}
@media only screen and (max-width: 600px) {
    .searchform-container {
        width: 98%;
    }
}