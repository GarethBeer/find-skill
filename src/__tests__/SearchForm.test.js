/* eslint-disable class-methods-use-this */
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import SearchForm from "../components/SearchForm";

describe("SearchForm", () => {
  const mockOnFill = jest.fn();
  const mockOnpick = jest.fn();

  const setUp = () => {
    const { asFragment, getByRole } = render(
      <SearchForm
        className="searchform"
        title="searchform"
        onFill={mockOnFill}
        onPick={mockOnpick}
      />
    );

    return {
      asFragment: asFragment(),
      input: getByRole("textbox", { name: /searchform/i }),
      largeRadio: getByRole("radio", { name: /large/i }),
      mediumRadio: getByRole("radio", { name: /medium/i }),
      smallRadio: getByRole("radio", { name: /small/i }),
      largeRadioLabel: getByRole("radio", { name: /large/i }).closest("label"),
      mediumRadioLabel: getByRole("radio", { name: /medium/i }).closest(
        "label"
      ),
      smallRadioLabel: getByRole("radio", { name: /small/i }).closest("label"),
    };
  };

  let renderedComponent;

  const setupGoogleMock = () => {
    const google = {
      maps: {
        places: {
          Autocomplete: class {
            setComponentRestrictions() {}

            setTypes() {}

            setFields() {}

            addListener() {}
          },
        },
      },
    };
    global.window.google = google;
  };

  beforeEach(() => {
    renderedComponent = setUp();
  });

  beforeAll(() => {
    setupGoogleMock();
  });

  it("renders correctly to match the SearchForm snapshot", () => {
    expect(renderedComponent.asFragment).toMatchSnapshot();
  });

  it("renders an properly", () => {
    expect(renderedComponent.input).toBeInTheDocument();
    expect(renderedComponent.input).toHaveClass("searchform");

    expect(renderedComponent.largeRadio).toBeInTheDocument();
    expect(renderedComponent.largeRadioLabel).toHaveTextContent(/large/i);
    expect(renderedComponent.largeRadioLabel).toHaveClass("radio");

    expect(renderedComponent.mediumRadioLabel).toHaveTextContent(/medium/i);
    expect(renderedComponent.mediumRadio).toBeInTheDocument();
    expect(renderedComponent.mediumRadioLabel).toHaveClass("radio");

    expect(renderedComponent.smallRadio).toBeInTheDocument();
    expect(renderedComponent.smallRadioLabel).toHaveTextContent(/small/i);
    expect(renderedComponent.smallRadioLabel).toHaveClass("radio");
  });

  it("searchform functions", () => {
    expect(renderedComponent.input.value).toBe("");
    fireEvent.change(renderedComponent.input, { target: { value: "address" } });
    expect(renderedComponent.input.value).toBe("address");
  });

  it("radio buttons function", () => {
    fireEvent.click(renderedComponent.largeRadio);
    expect(renderedComponent.largeRadio).toBeChecked();
    expect(renderedComponent.mediumRadio).not.toBeChecked();
    expect(renderedComponent.smallRadio).not.toBeChecked();

    fireEvent.click(renderedComponent.mediumRadio);
    expect(renderedComponent.largeRadio).not.toBeChecked();
    expect(renderedComponent.mediumRadio).toBeChecked();
    expect(renderedComponent.smallRadio).not.toBeChecked();
    expect(mockOnpick).toHaveBeenCalled();

    fireEvent.click(renderedComponent.smallRadio);
    expect(renderedComponent.largeRadio).not.toBeChecked();
    expect(renderedComponent.mediumRadio).not.toBeChecked();
    expect(renderedComponent.smallRadio).toBeChecked();
    expect(mockOnpick).toHaveBeenCalled();
  });
});
