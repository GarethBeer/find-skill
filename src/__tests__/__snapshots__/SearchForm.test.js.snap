// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders correctly to match the SearchForm snapshot 1`] = `
<DocumentFragment>
  <div
    class="searchform-container"
  >
    <h4>
      Pick your town
    </h4>
    <input
      class="searchform"
      placeholder="Enter a City"
      title="searchform"
    />
    <h4>
      Pick the search radius
    </h4>
    <form
      class="radius-container"
    >
      <label
        class="radio"
        for="8"
      >
        <input
          id="8"
          type="radio"
          value="8"
        />
        Large
      </label>
      <label
        class="radio"
        for="9"
      >
        <input
          id="9"
          type="radio"
          value="9"
        />
        Medium
      </label>
      <label
        class="radio"
        for="12"
      >
        <input
          id="12"
          type="radio"
          value="12"
        />
        Small
      </label>
    </form>
  </div>
</DocumentFragment>
`;
